### ISSUE TITLE
how to debug unwanted screen clearing

### ISSUE THREAD

**Issue by archenemies** [urwid/urwid#396]
##### Description:

I am trying to debug a program that uses urwid, namely Alot (see [issue 1469 in alot](https://github.com/pazz/alot/issues/1469)).

This program clears the screen when it exits. The author says it is standard behavior for urwid. I don't like the screen to be cleared, because I have a hard time remembering what was on it before it was cleared.

I don't really know if this is a bug or a feature request or just a request for help.

I noticed in `/usr/lib/python3.8/site-packages/urwid/curses_display.py`, there is a call to:

    curses.endwin()

However, the man page for `endwin` only talks about clearing the line.

Here is the final sequence, as reported by `script`, sent by Alot. I think somewhere in here the screen is getting cleared:

    ^H^[[0;1;37;1;48;5;17m^O^[[4he^[[4l^[[?25l^[[0;39;49m^[[H^[[1;1H^[[?1002l^[[?1000l^[[0;39;49m^O^[[?47l^[8^[[?25h^[]0;

I don't know how to decode these sequences, maybe I can get my terminal to print debugging information as it interprets them.

I noticed that Neomutt clears the screen, but Mutt does not. [Here](https://mutt-users.mutt.narkive.com/ugfB4p7e/exit-mutt-with-clear-screen) is an old thread which talks about that phenomenon. I try to disable secondary screen support in my terminal, as well as setting `smcup` and `rmcup` to the empty string in my terminfo file.

Many thanks if someone can point me forward on this.

##### Affected versions (if applicable)

Most recent Arch package.

**Comment by archenemies** [comment]
I found that the screen clearing is done by the following escape sequences:

    ^[[H
    ^[[1;1H

Both seem to have the same effect, although only the first is to be found in my terminfo file. It is under `clear` and `home`.

I think sending the sequences `smcup` and `rmcup` would be better because they seem to relate to the alternate screen, which can be disabled at the terminal level. Also, by using these sequences, users who prefer to have the alternate/secondary screen enabled will get to see the shell session restored which happened before running the Urwid client program, rather than having their screen cleared.

See here: <https://stackoverflow.com/questions/11023929/using-the-alternate-screen-in-a-bash-script>

However, I am confused because looking at the code for Urwid, it seems to use Curses, which as I have verified does not clear the screen when `endwin` is called (I compiled the example C program at the above link).

**Comment by archenemies** [comment]
This change fixes the problem on a test script:

    diff --git a/urwid/raw_display.py b/urwid/raw_display.py
    index f07c50d..1598621 100644
    --- a/urwid/raw_display.py
    +++ b/urwid/raw_display.py
    @@ -260,7 +260,7 @@ class Screen(BaseScreen, RealTerminal):

             move_cursor = ""
             if self._alternate_buffer:
    -            move_cursor = escape.RESTORE_NORMAL_BUFFER
    +            move_cursor = "" #escape.RESTORE_NORMAL_BUFFER
             elif self.maxrow is not None:
                 move_cursor = escape.set_cursor_position(
                     0, self.maxrow)

Here is the script:

    $ cat s.py
    import urwid

    def show_or_exit(key):
        if key in ('q', 'Q'):
            raise urwid.ExitMainLoop()
        txt.set_text(repr(key))

    txt = urwid.Text(u"Hello World\nHello World\nHello World\nHello World\n")
    fill = urwid.Filler(txt, 'top')
    loop = urwid.MainLoop(fill, unhandled_input=show_or_exit)
    loop.run()

I noticed that the sequences Urwid sends are hardcoded, and not in my terminfo database.

**Comment by ulidtko** [comment]
Hi!

One entry point to escape sequences is [`man console_codes`](http://man7.org/linux/man-pages/man4/console_codes.4.html). There're *decades* of legacy, standards, xterm emulation and whatnot in there — so beware.

Next, `raw_display` and `curses_display` are sort of alternate "rendering backends" in Urwid (an app can choose which to use). Expect terminfo usage in `curses_display`; expect hardcoded sequences in `raw_display`.

From pazz/alot#1469 I understand the issue you have: instead of clearing the alternate screen, urwid switches back to primary screen and clears that. Is that right?

**Comment by archenemies** [comment]
Let me get back to you on Monday, but I don't think that's quite right. I think Alot is doing something extra that Urwid isn't doing, so they are broken in different ways. The main thing of course is that the screen shouldn't get cleared when you have `XTerm*titeInhibit: false`

**Comment by ahmedkhalf** [comment]
> I understand the issue you have: instead of clearing the alternate screen, urwid switches back to primary screen and clears that. Is that right?

@ulidtko I am also having the same issue recently. Interestingly however I cannot reproduce it inside of konsole, but this issue happens with alacritty (with any urwid program). Furthermore, I was able to get a "hackish" workaround by using the terminal library blessed:
```python
import sys
import urwid

def show_or_exit(key):
    if key in ('q', 'Q'):
        raise urwid.ExitMainLoop()
    txt.set_text(repr(key))

txt = urwid.Text(u"Hello World")
fill = urwid.Filler(txt, 'top')
loop = urwid.MainLoop(fill, unhandled_input=show_or_exit)

import blessed
with blessed.Terminal().fullscreen():
    loop.run()

```
But it would be great to have this problem fixed from source rather bloating the software with another library.

**Comment by ulidtko** [comment]
Hey @ahmedkhalf, great finding. Can confirm the issue repro's on Alacritty 0.7.1, the "primary" screen gets cleared on exit. But it doesn't on libVTE-based terminals (`gnome-terminal`, Guake, etc) — similarly to what you found with `konsole`.

From htop/vim/less, it's evident that alacritty does implement the "alternate" screen feature, there's no issue with the emulator itself.

Using `script` I've compared the escape sequences from OP's `test-vanilla.py` and your `test-blessed.py`:

| | |
|------|-----|
| `vanilla` | `^[[K^[[?1002l^[[?1000l^[[0;39;49m^O^[[?47l^[8^[[?25h^[[1m^[[7m%^[[27m^[[1m^[[0m` |
| `blessed` | `^[[K^[[?1002l^[[?1000l^[[0;39;49m^O^[[?47l^[8^[[?25h^[[?1049l^[[23;0;0t^[[1m^[[7m%^[[27m^[[1m^[[0m` |

So, on Alacritty, `blessed` does something extra: **`^[[?1049l^[[23;0;0t`**.

I couldn't quickly decode that. The [terminfo entry][] for Alacritty wasn't enlightening either. Ideas?

[terminfo entry]: https://invisible-island.net/ncurses/terminfo.ti.html#toc-_Alacritty

**Comment by archenemies** [comment]
It looks like I promised a reply almost a year ago and then forgot, sorry about that.

I'm not sure what is being debugged at the moment. `?1049l` is rmcup in my terminfo, I don't know about the rest.

It seems that it should be easy in Python to redirect stdout to a class that after each character prints a stack trace to a log file, then we could find what is printing the various escape sequences. 'from contextlib import redirect_stdout'?

Out of curiosity, what was wrong with my 3 March 2020 patch which avoids printing `escape.RESTORE_NORMAL_BUFFER` in `Screen._stop`? Were we waiting for a version that does something more sensible?

**Comment by ulidtko** [comment]
@archenemies thanks for getting back!

> I'm not sure what is being debugged at the moment.

Can we close this issue then?

> Out of curiosity, what was wrong with my 3 March 2020 patch which avoids printing escape.RESTORE_NORMAL_BUFFER in Screen._stop?

Sorry, but it's a stretch to call that a patch. Multiple reasons, I'll articulate 2:
 * It's incomplete.
   The "exit alternate screen" (a.k.a. `rmcup`) sequence is removed, but  `escape.SWITCH_TO_ALTERNATE_BUFFER` (a.k.a. `smcup`) is not. It seems you didn't test this, but "the patch" will obviously make a mess on terminals which support and enable the alternate screen — because urwid will mutate the term's state and never restore it.
   For example, can your terminal **scroll back** after running a sample with "the patch" applied? I bet it can't — because it's stuck in the wrong mode — because the "patch" breaks the proper termination of alternate screen mode.

 * It's wrong, in the aspect of intention.
   As I said, *Expect terminfo usage in `curses_display`; expect hardcoded sequences in `raw_display`.* 

Hopefully that's enough for your curiousity. I mean no offense, it's just very far from a solution, despite perhaps appearing like it is.

--------

Meanwhile, I just saw that `tput rmcup` produces exactly the *blessed* sequence, `^[[?1049l^[[23;0;0t` — on both libvte & alacritty.

`^[[22;0;0t` and `^[[23;0;0t` are instances of [`DECRARA`](https://vt100.net/docs/vt510-rm/DECRARA.html), the purpose of which as part of `smcup`/`rmcup` is puzzling to me :thinking:

The `^[[?1049h`/`^[[?1049l` are [xterm's extensions][xterm] to the "DEC Private Mode Set". Quoting from the doc:

> P s = 1049 → Save cursor as in DECSC and use Alternate Screen Buffer, clearing it first (unless disabled by the `titeInhibit` resource). This combines the effects of the 1047 and 1048 modes. Use this with terminfo-based applications rather than the 47 mode.

Which is pretty enlightening, at least at the XTerm level of the compatibility tower. Maybe something else is going on at higher levels of emulation, but `\x1b[?1049h` should have the same effect as `ESC 7 ESC [ ? 4 7 h` to enter the alternate screen mode, and conversely `\x1b[?1049l` should be equivalent to `ESC [ ? 4 7 l ESC 8` to return to "normal" mode. 

[xterm]: https://www.xfree86.org/current/ctlseqs.html

Urwid's raw_display does the latter, i.e. `^[[?47l^[8` — and thus Should Work™; and not only with xterm-compatible terminals, but with any VT100/VT220-compatible emulators, I believe. And in fact does work, as we've seen, at least konsole & libvte do handle that just fine.

I could imagine, from a terminal implementer's perspective, that the ambiguity (are DECSET/DECRST modes 1049 and 47 the same or distinct? i.e. do they share an `is_alternate_screen` boolean flag underneath?..) could've caused such a bug in a reimplemented-anew emulator such as Alacritty. Perhaps it's worthwhile to talk to them about this, of course after more thorough testing of Alacritty specifically.

On Urwid side — at this point I don't think we should change anything.

**Comment by archenemies** [comment]
@ulidtko Sure thing.

1. What should I call the output of `diff`? I apologize if I was using the wrong term and didn't mean to imply anything inappropriate.

2. Do you want all Urwid applications to clear the screen on exit, or is it OK to let the user configure that using `XTerm*titeInhibit: true`? Because if you don't want it to be configurable, then there's not much point in my testing the modification I proposed to make it configurable.

3. What should I test? Just let me know. I have limited knowledge of your library. Partly that is because I avoid applications that use it, because of this issue.

    I tried this:

        xterm -xrm "XTerm*titeInhibit: false"

    and then I ran a program

        python urwid-test.py

    and checked that it restores the "alternate screen" on exit, just as say `mutt` does in the same terminal. Then I ran:

        xterm -xrm "XTerm*titeInhibit: true"

    and checked that the screen is not cleared on exit for both `urwid-test.py` and `mutt`. Scrolling works fine in both cases.

    This is the desired behavior for me, and it results from using a local Urwid installation which has been modified using the instructions I produced via the `git diff` command in my earlier comment. If I use the Urwid version from my distribution repository, then the screen is always cleared on exit (although with `urwid-test.py`, it leaves behind one line). This clearing is done whether or not `XTerm*titeInhibit` is `true`, and is undesired for me.

    Here is `urwid-test.py`:

        $ cat urwid-test.py
        import urwid

        def show_or_exit(key):
            if key in ('q', 'Q'):
                raise urwid.ExitMainLoop()
            txt.set_text(repr(key))

        txt = urwid.Text(u"Hello World\nHello World\nHello World\nHello World\n")
        fill = urwid.Filler(txt, 'top')
        loop = urwid.MainLoop(fill, unhandled_input=show_or_exit)
        loop.run()

4. I don't think you need my permission to close the issue, if that is your goal.

**Comment by ulidtko** [comment]
@archenemies great, thanks for complete self-contained repro case.

My results with it are different.

With `xterm -xrm "XTerm*titeInhibit: false"` — Screen **not cleared**, scrolling works.

With `xterm -xrm "XTerm*titeInhibit: true"` — Only last line remains, previous buffer is scrollable, but partially overwritten.

This is after running your exact `urwid-test.py` and exiting from it; `xterm -version` says XTerm(366).

I don't have urwid installed on this machine in any way; `pip list | grep urwid` outputs nothing, and `cd /tmp && python -c 'import urwid'` produces `ModuleNotFoundError`. Please make sure you likewise don't have any old/stale/unexpected installation of urwid lingering somewhere — that's the only possible explanation I see for the difference in our test's results.

> If I use the Urwid version from my distribution repository

Please specify which version that is exactly. I've no way of knowing what your distro has. `python -c 'import urwid; print(urwid.VERSION)'` is one way to get it. On my machine, this outputs `(2, 1, 3, 'dev')` — because as I said, I don't have it installed and am testing directly from git (current master, commit `c93aeb23fbca0ce8f12187ebc3ca330443a3bf6b` which is `release-2.1.2` + 11 commits).

Learning about your Python version could help as well.

**Comment by ulidtko** [comment]
I also don't reproduce the screen-clearing behavior in:
 * urxvt v9.22
 * `gnome-terminal` 3.38.3 using VTE 0.62.3 +BIDI +GNUTLS +ICU +SYSTEMD
 * Guake 3.7.0

The only terminal where the main scrollback buffer *does* get cleared during this test, leaving the 4 `Hello World` lines visible on screen — is Alacritty 0.7.1, but that looks like a separate issue.

**Comment by archenemies** [comment]
Thanks @ulidtko.

* You said "The only terminal where the main scrollback buffer does get cleared during this test, leaving the 4 Hello World lines visible on screen" ... I think maybe we have a different definition of "clearing"? If the screen is cleared at exit, the Hello World lines disappear. What I am complaining about is the fact that the text displayed by the Urwid client (here `urwid-test.py`) is erased when Urwid exits. I would like the text to remain on the screen. The text is erased because Urwid sends a terminal escape sequence which clears the screen, so for me it's just a matter of not sending this sequence. With a typical Curses application, disabling the alternate screen (`XTerm*titeInhibit: true`) is sufficient to make the text stay on the screen.

* Modulo the above, what behavior are we going for? Do you agree with me that screen clearing should be configurable?

* I don't understand how you can run `urwid-test.py` without Urwid installed. When I try that I just get an error

      $ python ./urwid-test.py
      Traceback (most recent call last):
        File "./urwid-test.py", line 1, in <module>
          import urwid
      ImportError: No module named urwid

* When you ran `xterm`, did you try running `mutt` or some other Curses application to see if the screen is cleared on exit? As I said, with `xterm -xrm "XTerm*titeInhibit: true"`, Mutt's screen stays visible after exit, while with `xterm -xrm "XTerm*titeInhibit: false"`, the text goes away (original screen with shell interaction is restored).

* Can you remind me how to install a local version of Urwid from this repository? In the `README.rst` you recommend `pip install urwid`, but as far as I can tell this is installing from a different source. I must have installed a version with my own modifications when I first submitted this issue. Is this the recommended syntax?

      python setup.py install --prefix ~/.local

  but it looks like I also have to delete the `build` directory to get my changes to be installed?

* Do I need to run this command whenever I upgrade Python? I noticed that the Python version number is part of the installation path:

      .local/lib/python3.9/site-packages/urwid-2.1.1.dev0-py3.9-linux-x86_64.egg/urwid

You may be right that I have been loading a "stale" version of Urwid, because now it is as you said, that my modified version leaves text on the screen under `xterm -xrm "XTerm*titeInhibit: false"`, and the scrollback doesn't work. The system version of Urwid shows the same behavior as Mutt, restoring the screen on exit. I ran `git pull` so now it is Urwid version 2.1.3.dev0. I apologize for any wasted time.

How about this "git-diff output"? It seems to produce the behavior I described in my previous comment, restoring the alternate screen conditional on `titeInhibit: false`, otherwise leaving text on the screen. Although maybe not a complete solution since it only behaves like Mutt when the cursor is on the last line, otherwise there will be a number of blank lines when the screen is restored. Do you know what `printf '\0338'` does?

    diff --git a/urwid/escape.py b/urwid/escape.py
    index be9cfd0..127ecd7 100644
    --- a/urwid/escape.py
    +++ b/urwid/escape.py
    @@ -410,7 +410,8 @@ APP_KEYPAD_MODE = ESC+"="
     NUM_KEYPAD_MODE = ESC+">"

     SWITCH_TO_ALTERNATE_BUFFER = ESC+"7"+ESC+"[?47h"
    -RESTORE_NORMAL_BUFFER = ESC+"[?47l"+ESC+"8"
    +#RESTORE_NORMAL_BUFFER = ESC+"[?47l"+ESC+"8"
    +RESTORE_NORMAL_BUFFER = ESC+"[?47l"

     #RESET_SCROLL_REGION = ESC+"[;r"
     #RESET = ESC+"c"

Also, the above only works on Xterm, with Urxvt it still clears the screen. I wonder how much can be done with hard-coded escape sequences.
 
Here is some version information.

    $ xterm -v
    XTerm(366)
    $ python --version
    Python 3.9.1
    $ python -c 'import urwid; print(urwid.VERSION)'
    (2, 1, 3, 'dev')

**Comment by archenemies** [comment]
Not sure if this helps, but here are files that should contain the initialization and de-initialization sequences emitted by Mutt on Urxvt and Xterm respectively, on my system:

[test-urxvt.txt](https://github.com/urwid/urwid/files/6032135/test-urxvt.txt)
[test-xterm.txt](https://github.com/urwid/urwid/files/6032136/test-xterm.txt)

I'm not sure what the sequences do, but I can see that they are different.

**Comment by ulidtko** [comment]
> I think maybe we have a different definition of "clearing"? If the screen is cleared at exit, the Hello World lines disappear.

Yes. Notice that *"the screen"* is ambiguous — maybe that's why we're getting confused. There're normally *two screens* (more precisely, two screen buffers) per terminal. Notice how I say "main scrollback buffer" or "alternate screen buffer" to be unambiguous.

Programs like `man`, `less`, `vim`, `htop`, `mutt`, Midnight Commander — all do their drawing in the *alternate* screen buffer. Urwid is no different. The alternate buffer has no scrolling; only the "main buffer" preserves history which you can scroll back to. Maybe see [this question](https://stackoverflow.com/questions/11023929/using-the-alternate-screen-in-a-bash-script) or other sources which explain the concept better.

Next, obviously, the terminal cannot display both buffers at once. Only one is visible at a time. Hence, terminals have a *mode*, that is the Alternate Screen Mode. There are escape sequences, let's just call them `smcup` and `rmcup` — those are the commands for the terminal to enter and exit the Alternate Screen Mode. After `smcup`, what you see in the terminal is the Alternate Screen Buffer; after `rmcup` what you see is the Main Screen Buffer again.

Next, not all terminals *support* the Alternate Screen Mode feature. Most modern ones do; but there's a lot of weirdness in the terminal-emulation space... some don't. Enabling the xterm's [`titeInhibit`][titeInhibit] preference (X11 lingo calls those *resources*) has the effect of disabling the Alternate Screen feature, and pretending it's not supported in xterm. With `titeInhibit` set, the terminal is always using just a single main buffer.

It's your preference to pick, but I wouldn't generally recommend enabling it. You'll start getting all sorts of log corruption, parts of output getting overwritten, color modes not getting reset and spilling over, mouse scrolling and arrow keys spewing garbage... You know, all those terminal state desynchronization issues usually cured by `tput reset`.

The whole Alternate Screen feature likely had reasons to get invented.

--------------------

Having that, let me just highlight that here I was speaking of clearing the *main scrollback buffer*:
> The only terminal where the main scrollback buffer *does* get cleared during this test, leaving the 4 Hello World lines visible on screen — is Alacritty 0.7.1, but that looks like a separate issue.

Steps:
 * launch alacritty
 * `ls /usr/share/doc` to generate filler output in the main/primary buffer
 * `cd ~/src/urwid`
 * `python urwid-test.py`
 * <kbd>q</kbd>
 * observe the previous filler output is gone.

We need to be careful of which of the buffers gets cleared.

------------

> What I am complaining about is the fact that the text displayed by the Urwid client (here urwid-test.py) is erased when Urwid exits. I would like the text to remain on the screen.

Understood. You want something like `less -X /proc/uptime` does. 

------------

> Modulo the above, what behavior are we going for?

Generally speaking, as the issue reporter you are expected to formulate the "expected behavior" (and then contrast it with "actual behavior" from a Repro Steps sequence, demonstrating a possibly-bug). See e.g. here: https://rubygarage.org/blog/how-to-write-a-quality-bug-report

Don't take me wrong; this isn't the best bug report in the world, but still a very decent discussion of a complex issue. I'm enjoying it and all the learning required to do it.

> Do you agree with me that screen clearing should be configurable?

It's somehow difficult for me to answer "yes" in context; yet I can easily disagree to the opposite ("Should screen clearing be not configurable?" — I answer "disagreed"). I know it sounds ridiculous; it feels ridiculous too. Sorry. Perhaps the issue is that I see too many possible points of confusion at play, and don't want my agreement misinterpreted.

E.g. I'm still not sure if you include enabling `titeInhibit` into the meaning of "configuring screen clearing", are you? That's a problem, because screen clearing is not at all the primary purpose of `titeInhibit`, not at all. It's only a weird distant side-effect.

What I wrote above should provide more clarity in this question.

--------------
> I don't understand how you can run urwid-test.py without Urwid installed.

Make sure to run from topdir of a git checkout:

    $ cd ~/src
    $ git clone https://github.com/urwid/urwid
    $ cd urwid
    $ cp .../urwid-test.py .
    $ python urwid-test.py

>  Is this the recommended syntax?
>
>       python setup.py install --prefix ~/.local

No. Try `pip install --user -e .` from the git topdir. The `-e` will make any changes in git immediately visible without reinstall; this way you can avoid extra unexpected copies. Variations of `setup.py install` will work too, but I'd recommend using `pip`; it can uninstall packages, `pip list`, `pip show urwid`, etc.

At least on Ubuntu, you have to `apt install python3-setuptools python3-pip` once, before `pip` works.

>  Do I need to run this command whenever I upgrade Python? I noticed that the Python version number is part of the installation path:
>
>     .local/lib/python3.9/site-packages/urwid-2.1.1.dev0-py3.9-linux-x86_64.egg/urwid

Yes that's correct, all installed packages are specific to the Python version. After even a minor update, you need to `pip install` all the things again.

--------------
> [...] with `xterm -xrm "XTerm*titeInhibit: false"`, the text goes away (original screen with shell interaction is restored)

That's exactly the expected behavior. The terminal switches off the alternate screen, and then the main screen buffer is active.

>  with `xterm -xrm "XTerm*titeInhibit: true"`, Mutt's screen stays visible after exit

Yes. But you also get part of the main buffer erased (overwritten) with Mutt. This should also be expected once you realize what `titeInhibit` does.

--------------

> Do you know what printf '\0338' does?

This is `DECRC`, a sequence of 2 characters:
 * `\033`, the U+001B `ESC` character in octal notation,
 * `8` the usual ascii digit.

If you look carefully in [`man console_codes`][man] or the [xterm documentation][], you'll find what it does:

>       ESC 7     DECSC    Save current state (cursor coordinates, attributes, character sets pointed at by G0, G1).
>       ESC 8     DECRC    Restore state most recently saved by ESC 7.

So,

```diff
-RESTORE_NORMAL_BUFFER = ESC+"[?47l"+ESC+"8"
+RESTORE_NORMAL_BUFFER = ESC+"[?47l"
```
— still no good, I hope you can see. Not restoring cursor position & colors ("attributes") will lead to ugly issues.

> What should I call the output of diff?

A diff, a change, a patch :) My apologies too, as I was partly wrong claiming it's not a patch. I simply don't know a better word for a logically complete, correct, self-contained *mergeable* change. Let's not dwell on that.

----------------

> Here is some version information.

Perfect :+1: My versions are the same; you're using Arch too?

Now, I got a bit lost; we've established matching environment — do we also have test behaviors in agreement?

---------------

> the above only works on Xterm, with Urxvt it still clears the screen

Interesting. What I think might be happening here, perhaps the different terminals interpret `rmcup` differently: some wipe the alternate screen buffer, some preserve it. Digging the source of urxvt & xterm might be the only guaranteed strategy to get to the bottom of it.

The dumps you posted do help. There I can see the so-called "ECMA-48 CSI" sequences:
 * `\033[nn;mmH` — `CUP` — Move cursor to row `nn` col `mm`
 * `\033[K` — `EL` — Erase line
 * `\033[J` — `ED` — Erase display

Strangely enough, the `ESC [ J` doesn't appear in my runs. E.g. here, if you can already read those, the sequence iterates over all the 38 lines, erases each one using colors reset `x1b[0;39;49m` + `\x0f` + `EL`, and finally turns off [mouse tracking][xterm documentation] with `\x1b[?1002l\x1b[?1000l`, exits alternate screen with `\x1b[?47l`, restores state with `DECRC` and turns cursor visible with `DECTECM`=`ESC [ ? 25 h`:

    $ script -q -c 'python urwid-test.py' /tmp/term.dump
    $ python -c 'print(repr(open("/tmp/term.dump").read()))'

<code>
'Script started on 2021-02-24 19:21:34+02:00 [TERM="xterm-256color" TTY="/dev/pts/7" COLUMNS="211" LINES="38"]\n\x1b7\x1b[?47h\x1b[?1002l\x1b[?1000l\x1b[?1000h\x1b[?1002h\x1b)0\x1b[?25l\x1b[0;39;49m\x1b[H\x1b[1;1H\x1b[0;39;49m\x0fHello World\x1b[K\x1b[2;1H\x1b[0;39;49m\x0fHello World\x1b[K\x1b[3;1H\x1b[0;39;49m\x0fHello World\x1b[K\x1b[4;1H\x1b[0;39;49m\x0fHello World\x1b[K\x1b[5;1H\x1b[0;39;49m\x0f\x1b[K\x1b[6;1H\x1b[0;39;49m\x0f\x1b[K\x1b[7;1H\x1b[0;39;49m\x0f\x1b[K\x1b[8;1H\x1b[0;39;49m\x0f\x1b[K\x1b[9;1H\x1b[0;39;49m\x0f\x1b[K\x1b[10;1H\x1b[0;39;49m\x0f\x1b[K\x1b[11;1H\x1b[0;39;49m\x0f\x1b[K\x1b[12;1H\x1b[0;39;49m\x0f\x1b[K\x1b[13;1H\x1b[0;39;49m\x0f\x1b[K\x1b[14;1H\x1b[0;39;49m\x0f\x1b[K\x1b[15;1H\x1b[0;39;49m\x0f\x1b[K\x1b[16;1H\x1b[0;39;49m\x0f\x1b[K\x1b[17;1H\x1b[0;39;49m\x0f\x1b[K\x1b[18;1H\x1b[0;39;49m\x0f\x1b[K\x1b[19;1H\x1b[0;39;49m\x0f\x1b[K\x1b[20;1H\x1b[0;39;49m\x0f\x1b[K\x1b[21;1H\x1b[0;39;49m\x0f\x1b[K\x1b[22;1H\x1b[0;39;49m\x0f\x1b[K\x1b[23;1H\x1b[0;39;49m\x0f\x1b[K\x1b[24;1H\x1b[0;39;49m\x0f\x1b[K\x1b[25;1H\x1b[0;39;49m\x0f\x1b[K\x1b[26;1H\x1b[0;39;49m\x0f\x1b[K\x1b[27;1H\x1b[0;39;49m\x0f\x1b[K\x1b[28;1H\x1b[0;39;49m\x0f\x1b[K\x1b[29;1H\x1b[0;39;49m\x0f\x1b[K\x1b[30;1H\x1b[0;39;49m\x0f\x1b[K\x1b[31;1H\x1b[0;39;49m\x0f\x1b[K\x1b[32;1H\x1b[0;39;49m\x0f\x1b[K\x1b[33;1H\x1b[0;39;49m\x0f\x1b[K\x1b[34;1H\x1b[0;39;49m\x0f\x1b[K\x1b[35;1H\x1b[0;39;49m\x0f\x1b[K\x1b[36;1H\x1b[0;39;49m\x0f\x1b[K\x1b[37;1H\x1b[0;39;49m\x0f\x1b[K\x1b[38;1H\x1b[0;39;49m\x0f\x1b[K\x1b[?1002l\x1b[?1000l\x1b[0;39;49m\x0f\x1b[?47l\x1b8\x1b[?25h\nScript done on 2021-02-24 19:21:35+02:00 [COMMAND_EXIT_CODE="0"]\n'</code>

-------------------

One more thing. I've never done it, but I'm pretty sure one can *configure* the terminfo database (`/usr/share/terminfo/*/*`, `man terminfo`) and modify what `tput rmcup` outputs on each terminal (there're separate entries for xterm, urxvt, alacritty, etc). It is strictly an `ncurses`-specific thing though; it will definitely affect urwid's `curses_display`, but `raw_display` — I'm not sure.

[man]: https://www.man7.org/linux/man-pages/man4/console_codes.4.html
[xterm documentation]: https://www.xfree86.org/current/ctlseqs.html
[titeInhibit]: https://invisible-island.net/xterm/manpage/xterm.html#VT100-Widget-Resources:titeInhibit

**Comment by archenemies** [comment]
@ulidtko Thank you for the comprehensive explanations.

I'm especially glad to know about `pip install --user -e .`. I was confused at first how it is able to tell Python where to look for the library. I thought it was going to install symlinks into `~/.local` but instead I see `~/.local/lib/python3.9/site-packages/{easy-install.pth,urwid.egg-link}` which I guess are something that Python understands.

Yes I'm using Arch.

> Now, I got a bit lost; we've established matching environment — do we also have test behaviors in agreement?

I'm not sure if there is reason to think so. It's possible that we were using different terminology or had different experiences but I think that my explorations and your explanations have made it clear that the problem can't be solved in a terminal-independent manner by emitting fixed escape sequences, which seems to make it all slightly irrelevant. If you can make fixed sequences work, then it would be great.

But realize I don't fully understand the difference between `raw_display` and `curses_display`. What controls which of these interfaces is used? My test script doesn't seem to specify it.

And `curses` comes with Python. Why is `alot` using the Urwid library rather than `curses`? Is it just that Urwid is simpler? I was under the impression that Urwid has been adopted by a lot of Python projects.

If `curses_display` relies on `curses`, does that mean that it checks Terminfo and so on like `curses` does?

Thanks.

**Comment by ulidtko** [comment]
@archenemies the display module to use is set via `urwid.MainLoop`'s [`screen` argument][screen arg]. This would be the modification to test with [`curses_display`](http://urwid.org/reference/display_modules.html#module-urwid.curses_display):
```diff
diff --git i/urwid-test.py w/urwid-test.py
index b0de3d2..d87fc40 100644
--- i/urwid-test.py
+++ w/urwid-test.py
@@ -1,4 +1,5 @@
 import urwid
+import urwid.curses_display
 
 def show_or_exit(key):
     if key in ('q', 'Q'):
@@ -7,5 +8,5 @@ def show_or_exit(key):
 
 txt = urwid.Text(u"Hello World\nHello World\nHello World\nHello World\n")
 fill = urwid.Filler(txt, 'top')
-loop = urwid.MainLoop(fill, unhandled_input=show_or_exit)
+loop = urwid.MainLoop(fill, unhandled_input=show_or_exit, screen=urwid.curses_display.Screen())
 loop.run()
```

[screen arg]: http://urwid.org/reference/main_loop.html#mainloop

Also check out how in [pudb][] they made the `raw_display`/`curses_display` a user-selectable preference item (pudb uses Urwid too).

> If curses_display relies on curses, does that mean that it checks Terminfo and so on like curses does?

Yes, exactly. You can check the source, `curses_display` does `import curses` and uses it directly; there're no hardcoded escape sequences in this implementation.

It has its drawbacks too, due to curses API limitations, e.g. this bug #412; also IIRC curses_display cannot do pipes/FDs in the main loop (you can call [`urwid.MainLoop.watch_pipe`](http://urwid.org/reference/main_loop.html#urwid.MainLoop.watch_file) but the callbacks will never fire).

[pudb]: https://pypi.org/project/pudb/

> Why is alot using the Urwid library rather than curses? Is it just that Urwid is simpler?

Urwid and curses are different layers of the stack, so to speak. By way of analogy, it's like Qt and X11: the former provides fancy widgets and rests on top of a windowing abstraction, the latter provides one implementation of that abstraction. As you can see, terminals are tricky, so having some of the complexity abstracted behind a "screen" interface is rather sane design. It also allows Urwid to have things like [an HTML "display"][1] and [`lcd_display`][2], though I think these are heavily underused and may need updates.

[1]: http://urwid.org/reference/display_modules.html#module-urwid.html_fragment
[2]: http://urwid.org/reference/display_modules.html#module-urwid.lcd_display

**Comment by archenemies** [comment]
@ulidtko Thank you for the clear explanations.

I checked that making the changes you proposed to `urwid-test.py` gives good behavior.

I tried modifying Alot (`alot/ui.py`) to use the Urwid Curses module according to the same diff, but when I do that, I get:

    NotImplementedError: screen object passed <urwid.curses_display.Screen object at 0x7f9126b11a00> does not support external event loops

So I can see that Alot depends on the `raw_display` module because of limitations in the Curses event API...?

In which case, progress in this direction is probably blocked by the Curses library.

And in order to fix this issue in `raw_display`, you would have to learn too much about Terminfo, or reimplement too much of the Curses escape sequence production logic, to make it worthwhile.

**Comment by neiljp** [comment]
To clarify the situation with this issue:

Do I understand correctly that this issue has been confirmed to concern removal (or enabling the configuration) of the use of the alternate screen buffer? (ie. using just one buffer, not two)
If so, may I suggest renaming this issue appropriately?

From discussion above, I assume by default the intentional behavior from urwid is that the alternate screen buffer should be used, and restored correctly on termination? As per the issue referenced by @ditsuke above (zulip/zulip-terminal#1152), we can reliably reproduce this not occurring in some terminals when using zulip-terminal (using urwid with raw backend). @ulidtko In analysis above you mentioned a separate problem, I think? Is that related to something like #487? The problem we've found could be how we're using urwid, but #487 appears to reproduce something similar in an example included in urwid?
(happy to continue this discussion there, once it's clear which issue concerns what elements!)

**Comment by neiljp** [comment]
Another follow-up, given this being notably reproducible with Alacritty, I noticed this closed issue: alacritty/alacritty#3869

I don't think that was touched on in the above discussion.

**Comment by AnonymouX47** [comment]
@penguinolog I think this issue should've also been resolved by #556, i just didn't notice it earlier.
